### PR TITLE
[SPARK-11468][SPARKR] add stddev/variance agg functions for Column

### PR DESCRIPTION
**Pull request by felixcheung**
Checked names, none of them should conflict with anything in base

@shivaram @davies @rxin

### CODE REVIEW

**Comment by shivaram** [comment]
cc @sun-rui

**Comment by davies** [comment]
LGTM

**Comment by mengxr** [comment]
Should we add aliases `var` and `sd` in SparkR, which are more familiar to R users?

**Comment by felixcheung** [comment]
@mengxr possibly.. though there are usage difference with SparkR DataFrame/Column as compared to  R data.frame (eg. `agg` vs [`aggregate`](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/aggregate.html), `var(faithful$eruptions)` would have been a Column with DataFrame)
It might be more confusing.

**Comment by shivaram** [comment]
Looks like dply has aggregate functions for variance, standard deviations ? Can we just match the syntax there ? https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf

**Comment by felixcheung** [comment]
This does work
```
> head(agg(df, stddev(df$age)))    # same as head(summarize(df, stddev(df$age)))
  stddev_samp(age)
1         7.778175
```
Also these work
```
> gd <- groupBy(df, "name")
> head(agg(gd, stddev(df$age)))    # same as head(agg(gd, age = "stddev"))
     name stddev_samp(age)
1    Andy                0
2 Michael               NA
3  Justin                0
```
I think they match dplyr, so I could name them `sd` and `var`, but just as I mention, doesn't work as one would expect on data.frame column vector.

**Comment by sun-rui** [comment]
@shivaram, I looked at dplyr package, there is no "sd", "var" defined in the package, seems the page is referencing "sd" "var" in R base.

Yes, stddev/variance agg functions we are adding have similar syntax as "sd" and "var" in R, as you can think of a Column as a vector of same type. So we'd better add aliases for them.

Note that var() in R computes sample variance (I think "sd" also uses sample variance, but not verified) instead of population variance.

File: R/pkg/NAMESPACE [hunk 1]
```diff
@@ -209,7 +210,11 @@ exportMethods("%in%",
               "sin",
               "sinh",
               "size",
+              "skewness",
               "soundex",
+              "stddev",
+              "stddev_samp",
+              "stddev_pop",
```
Comment (sun-rui) [inline R/pkg/NAMESPACE]: "stddev_pop" before "stddev_samp",

File: R/pkg/NAMESPACE [hunk 2]
```diff
@@ -228,6 +233,9 @@ exportMethods("%in%",
               "unhex",
               "unix_timestamp",
               "upper",
+              "variance",
+              "var_samp",
+              "var_pop",
```
Comment (sun-rui) [inline R/pkg/NAMESPACE]: "var_pop" before "var_samp",

File: R/pkg/R/functions.R [hunk 3]
```diff
@@ -974,6 +1006,54 @@ setMethod("soundex",
...
+#' Aggregate function: alias for \link{stddev_samp}
+#'
+#' @rdname stddev
+#' @name stddev
+#' @family agg_funcs
+#' @export
+#' @examples \dontrun{stddev(df$c)}
+setMethod("stddev",
+          signature(x = "Column"),
+          function(x) {
```
Comment (sun-rui) [inline R/pkg/R/functions.R]: call stddev_samp directly

File: R/pkg/R/functions.R [hunk 4]
```diff
@@ -1168,6 +1248,54 @@ setMethod("upper",
...
+#' Aggregate function: alias for \link{var_samp}.
+#'
+#' @rdname variance
+#' @name variance
+#' @family agg_funcs
+#' @export
+#' @examples \dontrun{variance(df$c)}
+setMethod("variance",
+          signature(x = "Column"),
+          function(x) {
```
Comment (sun-rui) [inline R/pkg/R/functions.R]: call var_samp() directly

File: R/pkg/inst/tests/test_sparkSQL.R [hunk 5]
```diff
@@ -997,20 +1004,42 @@ test_that("group by", {
...
+  expect_equal(3, count(mean(gd)))
+  expect_equal(3, count(max(gd)))
+  expect_equal(30, collect(max(gd))[1,2])
+  expect_equal(1, collect(count(gd))[1,2])
+
+  mockLines2 <- c("{\"name\":\"ID1\", \"value\": \"10\"}",
+                  "{\"name\":\"ID1\", \"value\": \"10\"}",
+                  "{\"name\":\"ID1\", \"value\": \"22\"}",
+                  "{\"name\":\"ID2\", \"value\": \"-3\"}")
+  jsonPath2 <- tempfile(pattern="sparkr-test", fileext=".tmp")
```
Comment (sun-rui) [inline R/pkg/inst/tests/test_sparkSQL.R]: delete this temp file at the end of the test case

**Comment by shivaram** [comment]
Yeah in that case lets add aliases for `sd` and `var` as well and match the population/sample behavior.

@felixcheung I think similar functions like `mean` don't match base R data frame but they are useful when used inside an expression and so its fine to have aliases for this.

File: R/pkg/R/functions.R (hunk 3 above)
Comment (felixcheung) [inline R/pkg/R/functions.R]: This is an alias on the Scala/Spark side

File: R/pkg/R/functions.R (hunk 4 above)
Comment (felixcheung) [inline R/pkg/R/functions.R]: Same here.

**Comment by sun-rui** [comment]
LGTM

**Comment by davies** [comment]
Merging this into master and 1.6 branch, thanks!

**Pull request closed without merging**